src/character.cpp

```diff
@@ -11268,6 +11256,56 @@ bool Character::use_charges_if_avail( const itype_id &it, int quantity )
     return false;
 }
 
+int Character::available_ups() const
+{
+    int available_charges = 0;
+    if( is_mounted() && mounted_creature.get()->has_flag( MF_RIDEABLE_MECH ) ) {
+        auto *mons = mounted_creature.get();
+        available_charges += mons->battery_item->ammo_remaining();
+    }
+    if( has_active_bionic( bio_ups ) ) {
+        available_charges += units::to_kilojoule( get_power_level() );
+    }
+    available_charges += charges_of( itype_UPS_off );
+    available_charges += charges_of( itype_adv_UPS_off );
+
+    return available_charges;
+}
+
+int Character::consume_ups( int qty )
+{
+    const int wanted_qty = qty;
+
+    if( qty != 0 && is_mounted() && mounted_creature.get()->has_flag( MF_RIDEABLE_MECH ) &&
+        mounted_creature.get()->battery_item ) {
+        auto *mons = mounted_creature.get();
+        int power_drain = std::min( mons->battery_item->ammo_remaining(), qty );
+        mons->use_mech_power( -power_drain );
+        qty -= std::min( qty, power_drain );
+    }
+
+    if( qty != 0 && has_power() && has_active_bionic( bio_ups ) ) {
+        int bio = std::min( units::to_kilojoule( get_power_level() ), qty );
+        mod_power_level( units::from_kilojoule( -bio ) );
+        qty -= std::min( qty, bio );
+    }
+
+    inventory inv = crafting_inventory( tripoint_zero, -1 );
```

Comment:
I don't have access to my local repo now, but from what I remember isn't crafting_inventory cache invalidated on change of turn or position?
Isn't it harmful to performance? I recall doing something similar in different place and it slowed the game to almost standstill.

Reply:
Maybe. I don't know much about it. But the impact shoulsn't be worse than it already is.

The `Character::consume_ups` replaces `Character::use_charges` for UPS use and is basically just a cut down version of it.